### PR TITLE
Performance/limit plagiarism comparisons

### PR DESCRIPTION
**Pull request by krusche**
Exercises with many students can lead to extreme performance problems, because millions of objects would need to be saved in the database. This does not make sense, because no now would view those plagiarism comparisons anyway. Also, it can have severe performance issues and even break the system completely. Therefore, this PR limits the saved and returned results to the 500 with the highest similarity.

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.

### Steps for Testing
1. Create many submissions (>40) for a modeling, text or programming exercise
2. Invoked the plagiarism check on the exercise detail page with the minimal similarity setting (50).
3. Make sure only up to 500 results are shown and the whole operation does not take too much time.

(Note that we have up to 2.200 submissions per exercise in courses on production, which could potentially lead to 2.200 * 2.200 / 2 = 2.420.000 comparisons). While the comparisons themselves are rather fast, saving the values in the database is time-consuming.

@iozbeyli helped me to track some performance metrics with 40 similar realistic modeling submissions
* Comparing plagiarisms took 64ms and led to 728 comparisons. Saving those 728 plagiarism comparisons in the database took around 10,4 s

### CODE REVIEW

**Review by maximilianbergmann — COMMENTED** [review]
Started testing on TS3. Created 46 Text Submissions. Unfortunately, I did not succeed in searching for plagiarisms. Error (zone.js:3382 GET https://vmbruegge60.in.tum.de/api/text-exercises/6117/check-plagiarism?similarityThreshold=50&minimumScore=0&minimumSize=0 500) - but it could also be that I am doing something wrong :-(

File: src/main/java/de/tum/in/www1/artemis/domain/plagiarism/PlagiarismComparison.java [hunk 1]
```diff
@@ -16,7 +18,7 @@
  */
 @Entity
 @Table(name = "plagiarism_comparison")
-public class PlagiarismComparison<E extends PlagiarismSubmissionElement> extends DomainObject {
+public class PlagiarismComparison<E extends PlagiarismSubmissionElement> extends DomainObject implements Comparable {
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/domain/plagiarism/PlagiarismComparison.java:21]:
```suggestion
public class PlagiarismComparison<E extends PlagiarismSubmissionElement> extends DomainObject implements Comparable<PlagiarismComparison<E>> {
```

File: src/main/java/de/tum/in/www1/artemis/domain/plagiarism/PlagiarismComparison.java [hunk 2]
```diff
@@ -125,4 +127,9 @@ public PlagiarismStatus getStatus() {
     public void setStatus(PlagiarismStatus status) {
         this.status = status;
     }
+
+    @Override
+    public int compareTo(@NotNull Object o) {
+        return Double.compare(similarity, ((PlagiarismComparison) o).similarity);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/domain/plagiarism/PlagiarismComparison.java:133]:
```suggestion
    public int compareTo(@NotNull PlagiarismComparison<E> o) {
        return Double.compare(similarity, o.similarity);
```

File: src/main/java/de/tum/in/www1/artemis/domain/plagiarism/PlagiarismComparison.java (hunk 1 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/domain/plagiarism/PlagiarismComparison.java:21]: Thanks for the hint, I was trying something similar but unfortunately failed

**Comment by krusche** [comment]
> Started testing on TS3. Created 46 Text Submissions. Unfortunately, I did not succeed in searching for plagiarisms. Error (zone.js:3382 GET https://vmbruegge60.in.tum.de/api/text-exercises/6117/check-plagiarism?similarityThreshold=50&minimumScore=0&minimumSize=0 500) - but it could also be that I am doing something wrong :-(

The internal server error on TS3 says:

```
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: jplag.ExitException: Not enough valid submissions! (found 1 valid submissions)
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]:         at jplag.JPlag.run(JPlag.java:232)
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]:         at de.tum.in.www1.artemis.service.plagiarism.TextPlagiarismDetectionService.checkPlagiarism(TextPlagiarismDetectionService.java:118)
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]:         at de.tum.in.www1.artemis.web.rest.TextExerciseResource.checkPlagiarism(TextExerciseResource.java:579)
```

```
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "33013-artemis_test_user_12.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "33015-artemis_test_user_10.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45304-artemis_test_user_1.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45305-artemis_test_user_2.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45306-artemis_test_user_3.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45307-artemis_test_user_4.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45308-artemis_test_user_5.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45309-artemis_test_user_20.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45310-artemis_test_user_21.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45311-artemis_test_user_22.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45312-artemis_test_user_23.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45313-artemis_test_user_24.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45314-artemis_test_user_25.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45315-artemis_test_user_26.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45316-artemis_test_user_27.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45317-artemis_test_user_28.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45318-artemis_test_user_29.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45319-artemis_test_user_30.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45320-artemis_test_user_31.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45321-artemis_test_user_32.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45322-artemis_test_user_33.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45323-artemis_test_user_34.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45324-artemis_test_user_35.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45325-artemis_test_user_36.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45326-artemis_test_user_37.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45327-artemis_test_user_38.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45328-artemis_test_user_39.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45329-artemis_test_user_40.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45330-artemis_test_user_41.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45331-artemis_test_user_42.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45332-artemis_test_user_43.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45333-artemis_test_user_44.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45334-artemis_test_user_45.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45335-artemis_test_user_46.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45336-artemis_test_user_47.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45337-artemis_test_user_48.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45338-artemis_test_user_49.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45339-artemis_test_user_50.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45340-artemis_test_user_51.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45341-artemis_test_user_52.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45342-artemis_test_user_53.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45343-artemis_test_user_54.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45344-artemis_test_user_55.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45345-artemis_test_user_56.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45346-artemis_test_user_57.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45347-artemis_test_user_58.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45348-artemis_test_user_59.txt" is too short!
Apr 18 14:00:28 vmbruegge60 artemis.sh[18835]: Submission "45349-artemis_test_user_60.txt" is too short!
```

You can define the minimum size in the user interface. I think this translates to the number of words in the JPlag options (and not characters). This might not be obvious though. 

image

I asked @philippbauch to mention this in the UI for text plagiarism checks. Probably, he forgot it.

Would be awesome if someone can help me to change this text for text exercises

**Comment by krusche** [comment]
Ok, I fixed it immediately, it was really easy :-)

**Review by NicolasUllmann — COMMENTED** [review]
Tested on Server 1 with 41 submissions. Works as expected.

**Review by willergeorg — COMMENTED** [review]
Tested on TS1 with 42 submissions on text exercise. Works as expected and did not take long.

**Review by maximilianbergmann — APPROVED** [review]
Thanks for your answer @krusche. I tested it again with longer submissions on TS3 - worked well. Took about 30 seconds.

**Review by willergeorg — APPROVED** [review]

**Review by NicolasUllmann — APPROVED** [review]

**Review by ivohashamov — APPROVED** [review]
Tested on TS1, results get displayed in less than a second with over 40 submissions of text exercises.

**Pull request merged**